File: ui/src/config/pagination.js
export const PAGE_SIZE = 10
export const PICKER_LIMIT = 500

export function withPageParams(params, { offset = 0, limit = PAGE_SIZE } = {}) {
  const p = params instanceof URLSearchParams
    ? new URLSearchParams(params)
    : new URLSearchParams(params || '')
  p.set('offset', String(offset))
  p.set('limit', String(limit))
  return p.toString()
}

File: logic/pagination.py
"""صفحه‌بندی مشترک لیست‌ها — پیش‌فرض ۱۰ رکورد."""

PAGE_SIZE = 10
MAX_PAGE_SIZE = 500
PICKER_LIMIT = 500


def parse_page(params, default_limit=PAGE_SIZE):
    try:
        offset = max(0, int(params.get("offset") or 0))
    except (TypeError, ValueError):
        offset = 0
    try:
        limit = int(params.get("limit") or default_limit)
    except (TypeError, ValueError):
        limit = default_limit
    limit = min(max(1, limit), MAX_PAGE_SIZE)
    return offset, limit


def paginate(qs, params, default_limit=PAGE_SIZE):
    """برش queryset و متادیتای total/offset/limit."""
    offset, limit = parse_page(params, default_limit)
    total = qs.count() if hasattr(qs, "count") else len(qs)
    page = list(qs[offset : offset + limit])
    return page, {"total": total, "offset": offset, "limit": limit}
